src/transactions/TransactionFrame.cpp

```diff
@@ -1130,10 +1161,6 @@ TransactionFrame::applyOperations(SignatureChecker& signatureChecker,
             outerMeta.pushOperationMetas(std::move(operationMetas));
             outerMeta.pushTxChangesAfter(std::move(changesAfter));
 #ifdef ENABLE_NEXT_PROTOCOL_VERSION_UNSAFE_FOR_PRODUCTION
-            // ContractEvents should only occur in InvokeHostFunctionOps which
-            // should only occur in txs with a single op. This is enforced in
-            // InvokeHostFunctionOpFrame::doCheckValid but re-assert here.
-            releaseAssertOrThrow(mEvents.empty() || mOperations.size() == 1);
```

Comment:
I think this is wrong: if any operation fails, no event should be emitted as they should be rolled back. (needs a new test for that as well)

Reply:
This code is in the `if (success)` block, so no events will be emitted on failure. Are you saying we should make sure no events were emitted by the host in the failure case?

Reply:
The assert was there as a reminder for us to deal with events properly.
I misread the code (thought it was in the for loop).

I *think* the things to consider instead are:
* even on failure, we want to emit diagnostic events (work you've been working on), but check (in test ) that no events are emitted
* on success (all ops are successful), do we need to break down events per operation in the meta or  leave it like this where it's a mashup of all ops? I guess `mEvents` could also be a vector indexed by operation

Reply:
- With the most recent changes to events, the diagnostics events get rolled back on failure as well. I was thinking this still made sense because downstream consumers like stellar-expert would only show the call stack on success, similar to how it doesn't show information on failed transactions. If we expect users to use these events to debug failures (and add more events specifically for failure cases), then this needs to be revisited.   
- The xdr already has the events as a vector of vectors, so this code just needs to be updated to use that.

Reply:
Wrt showing events on failure, it gets messier if a nested contract contract uses try_call. So A calls B using try_call, B emits an event, B fails, A swallows the error and succeeds. That B event should be rolled back because B failed, but if we emit it, downstream users might assume that it succeeded. I guess you could assume the call failed because the "function call" event would exist but the "return value" event would be missing, but this needs some more thought.

Reply:
I made the change to separate events by operations. This PR does not depend on the diagnostics work, so that can be figured out separately.

Reply:
changes look good now (wrt multi-op). I agree diagnostics can be done separately

Reply:
One thing missing is the negative case: multi-op smart tx that has a failure at op #2 should result in 0 events

Reply:
If there's a failure at op 2, `mEvents` would still be populated, but we wouldn't call `outerMeta.pushContractEvents`, so no events would be emitted. I don't think we need to check anything here.